### PR TITLE
Added error handling functionality in handle_step

### PR DESCRIPTION
**Pull request by niclew17**

### CODE REVIEW

File: quest_test/custom_errors/custom_error.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        self.code = code
+
+    def __str__(self):
+        return f"{self.message} (Error Code: {self.code})"
+
+    def get_code(self):
+        return self.code
+
+
+# quest_test/custom_errors/custom_error.py
```
Comment (byubean) [inline quest_test/custom_errors/custom_error.py:22]: remove comment?

File: quest_test/test_get_exception_class.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    with pytest.raises(exception_class) as exc_info:
+        raise exception_class("This is a custom error message")
+    assert str(exc_info.value) == "This is a custom error message"
+
+def test_weird_custom_exception():
+    exception_class = get_exception_class('quest_test.custom_errors.custom_error.WeirdCustomException')
+    with pytest.raises(exception_class) as exc_info:
+        raise exception_class("This is a weird custom error", 404)
+    assert str(exc_info.value) == "This is a weird custom error (Error Code: 404)"
+    assert exc_info.value.get_code() == 404
```
Comment (byubean) [inline quest_test/test_get_exception_class.py:35]: These need to be indented

File: src/quest/historian.py [hunk 3]
```diff
@@ -183,6 +184,12 @@ def get_function_name(func):
     else:  # Callable classes
         return func.__class__.__name__
 
+def get_exception_class(exception_type: str):
```
Comment (byubean) [inline src/quest/historian.py:187]: I think this can be `_get_exception_class`.

File: src/quest/historian.py [hunk 4]
```diff
@@ -198,6 +205,7 @@ def __init__(self, workflow_id: str, workflow: Callable, history: History):
         # These things need to be serialized
         self._history: History = history
 
+
```
Comment (byubean) [inline src/quest/historian.py:208]: Run the PyCharm formatter

File: src/quest/historian.py [hunk 5]
```diff
@@ -620,6 +630,11 @@ async def record_external_event(self, name, identity, action, *args, **kwargs):
         else:
             result = function(*args, **kwargs)
 
+        serializer = self._serializers.get(type(result))
+        if serializer:
+            result = serializer.to_json(result)
+
+
```
Comment (byubean) [inline src/quest/historian.py:637]: Leave this out

File: src/quest/historian.py [hunk 6]
```diff
@@ -645,6 +660,11 @@ async def _replay_external_event(self, record: ResourceAccessEvent):
             record['action']
         )(*record['args'], **record['kwargs'])
 
+        serializer = self._serializers.get(type(result))
+        if serializer:
+            result = serializer.from_json(record['result'])
+
+
```
Comment (byubean) [inline src/quest/historian.py:667]: Leave this out

File: quest_test/test_step_error.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    global foo_calls
+    foo_calls = foo_calls/0
+
+async def longer_workflow2(text):
+    text = await double(text)
+    try:
+        await add_foo2(text)
+    except ZeroDivisionError as e:
+        global foo_calls
+        foo_calls += 1
```
Comment (byubean) [inline quest_test/test_step_error.py:83]: why use a global variable here?

File: quest_test/test_step_error.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    assert result == 'abcabcfooabcabcfoo'
+    assert double_calls == 2
+    assert foo_calls == 1
+
+
+
+@step
+async def add_foo2(text):
+    global foo_calls
+    foo_calls = foo_calls/0
```
Comment (byubean) [inline quest_test/test_step_error.py:75]: why use the global variable and not simply `1 / 0`?

**Comment by niclew17** [comment]
I just made those changes on the latest commit. Thanks!

**Review by byubean — APPROVED** [review]

**Pull request merged**